modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/AbstractGridCoverage2DReader.java

```diff
@@ -360,10 +360,21 @@ protected Integer setReadParams(
 
         // //
         //
-        // default values for subsampling
+        // Resolution requested. I am here computing the resolution required by
+        // the user.
         //
-        // //
-        readP.setSourceSubsampling(1, 1, 0, 0);
+        double[] requestedRes =
+                getResolution(
+                        requestedEnvelope,
+                        requestedDim,
+                        getCoordinateReferenceSystem(coverageName));
+
+        // /////////////////////////////////////////////////////////////////////
+        // DECIMATION ON READING
+        // /////////////////////////////////////////////////////////////////////
+        decimationOnReadingControl(coverageName, imageChoice, readP, requestedRes);
```

Comment:
I'd move the call to decimationOnReadingControl below the check for requested res being being null (can see tha the method is null safe, but it's a pointless extra call if requestedRes is null)